kafkaconsumer/kafkaconsumer_integration_test.go

```diff
@@ -0,0 +1,458 @@
+
+	consumer1, err := Join("kafkaconsumer.TestFunctionalParallelConsumersWithInterruption", TopicSubscription("test.4"), zookeeper, config)
+	if err != nil {
+		t.Fatal(t)
+	}
+
+	consumer2, err := Join("kafkaconsumer.TestFunctionalParallelConsumersWithInterruption", TopicSubscription("test.4"), zookeeper, config)
+	if err != nil {
+		t.Fatal(t)
+	}
+
+	// We start consuming messages from both consumers in parallel.
+	// Both should acknowledge 25 messages.
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		ts.assertMessages(consumer1, 25)
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		ts.assertMessages(consumer2, 25)
+	}()
+
+	wg.Wait()
+
+	// Now, we close one consumer, which means that the second consumer will consume all other messages.
+	safeClose(t, consumer1)
+
+	ts.assertMessages(consumer2, 50)
+
+	// Now, close the second consumer
+	safeClose(t, consumer2)
+	ts.assertDrained(consumer2)
+
+	ts.assertOffsets()
+}
+
+////////////////////////////////////////////////////////////////////
+// Helper functions
+////////////////////////////////////////////////////////////////////
+
+func safeClose(t *testing.T, c io.Closer) {
+	if err := c.Close(); err != nil {
+		t.Error(err)
+	}
+}
+
+func newTestState(t *testing.T, group string) *testState {
+	return &testState{t: t, group: group, done: make(chan struct{})}
+}
+
+type testState struct {
+	t           *testing.T
+	group       string
+	c           sarama.Client
+	coord       *sarama.Broker
+	offsetTotal int64
+	topics      []string
+	produced    int64
+	consumed    int64
+	done        chan struct{}
+}
+
+func (ts *testState) close() {
+	if ts.c != nil {
+		safeClose(ts.t, ts.c)
+	}
+}
+
+func (ts *testState) client() sarama.Client {
+	if ts.c == nil {
+		config := sarama.NewConfig()
+		config.Producer.Partitioner = sarama.NewRoundRobinPartitioner
+
+		client, err := sarama.NewClient(kafkaBrokers, config)
+		if err != nil {
+			ts.t.Fatal("Failed to connect to Kafka:", err)
+		}
+
+		ts.c = client
+	}
+
+	return ts.c
+}
+
+func (ts *testState) coordinator() *sarama.Broker {
+	if ts.coord == nil {
+		coordinator, err := ts.client().Coordinator(ts.group)
+		if err != nil {
+			ts.t.Fatal(err)
+		}
+		ts.coord = coordinator
+	}
+
+	return ts.coord
+}
+
+func (ts *testState) prepareConsumer(topics []string) {
+	client := ts.client()
+	coordinator := ts.coordinator()
+
+	request := &sarama.OffsetCommitRequest{
+		Version:       1,
+		ConsumerGroup: ts.group,
+	}
+
+	ts.topics = topics
+	for _, topic := range topics {
+		partitions, err := client.Partitions(topic)
+		if err != nil {
+			ts.t.Fatal("Failed to retrieve list of partitions:", err)
+		}
+
+		for _, partition := range partitions {
+			offset, err := client.GetOffset(topic, partition, sarama.OffsetNewest)
+			if err != nil {
+				ts.t.Fatal("Failed to get latest offset for partition:", err)
+			}
+			ts.offsetTotal += offset - 1
+
+			request.AddBlock(topic, partition, offset-1, 0, "")
```

Comment:
Timestamp `0` means the beginning of Unix epoch. As a result all committed offsets are expired immediately. (in my setup Kafka kept them around for 1 minute). So you need to use `ReceiveTime` (-1) instead.